File: scripts/prebuild/data-pipeline/utils/url-classifier.ts
const INSTAGRAM_HOSTS = new Set([
  'instagram.com',
  'www.instagram.com',
  'instagr.am',
]);
const FACEBOOK_HOSTS = new Set([
  'facebook.com',
  'fb.com',
  'm.facebook.com',
  'www.facebook.com',
  'fb.me',
]);
const THREADS_HOSTS = new Set(['threads.net', 'www.threads.net']);

export interface SocialUrls {
  instagram_url: string | null;
  facebook_url: string | null;
  threads_url: string | null;
}

export function classifySocialUrl(url: string | null | undefined): SocialUrls {
  const result: SocialUrls = {
    instagram_url: null,
    facebook_url: null,
    threads_url: null,
  };

  if (!url) return result;

  try {
    const host = new URL(url).hostname.toLowerCase();

    if (INSTAGRAM_HOSTS.has(host)) result.instagram_url = url;
    else if (FACEBOOK_HOSTS.has(host)) result.facebook_url = url;
    else if (THREADS_HOSTS.has(host)) result.threads_url = url;
  } catch {
    // invalid URL — return nulls
  }

  return result;
}

File: backend/utils/url_classifier.py
from urllib.parse import urlparse

_INSTAGRAM_HOSTS = frozenset({"instagram.com", "www.instagram.com", "instagr.am"})
_FACEBOOK_HOSTS = frozenset(
    {"facebook.com", "fb.com", "m.facebook.com", "www.facebook.com", "fb.me"}
)
_THREADS_HOSTS = frozenset({"threads.net", "www.threads.net"})


def classify_social_url(url: str | None) -> dict[str, str | None]:
    """Classify a URL into its social platform type.

    Returns a dict with keys instagram_url, facebook_url, threads_url.
    At most one key will be non-None. All others will be None.
    """
    result: dict[str, str | None] = {
        "instagram_url": None,
        "facebook_url": None,
        "threads_url": None,
    }
    if not url:
        return result
    try:
        host = urlparse(url).netloc.lower()
    except Exception:
        return result
    if host in _INSTAGRAM_HOSTS:
        result["instagram_url"] = url
    elif host in _FACEBOOK_HOSTS:
        result["facebook_url"] = url
    elif host in _THREADS_HOSTS:
        result["threads_url"] = url
    return result
